### ISSUE TITLE
confusion about optimizer delay - working correctly? (Transformer Big)

### ISSUE THREAD

**Issue by kellymarchisio** [marian-nmt/marian#259]
I have two nearly-identical Transformer Big models that differ only in that `optimizer-delay=1` in the first and `optimizer-delay=4` in the second. Both have `--mini-batch-fit`, `--maxi-batch 1000`, and `maxi-batch-sort: src` .

Here are the first lines of training:
optimizer-delay=1 :
`Ep. 1 : Up. 500 : Sen. 31905 : Cost 9.98 : Time 622.73s : 1409.48 words/s : L.r. 3.1250e-06`

optimizer-delay=4
`Ep. 1 : Up. 500 : Sen. 31905 : Cost 9.98 : Time 622.62s : 1409.73 words/s : L.r. 3.1250e-06`

I'd have expected from [this chain](https://github.com/marian-nmt/marian-dev/issues/225) for Sen. in optimizer-delay=4 to be 4x what I saw in optimizer-delay=1. Furthermore, it appears my models are only seeing ~64 sentences per update, correct? This doesn't seem right --- right?

**Comment by emjotde** [comment]
Are you training on a single GPU? If yes, add --sync-sgd. That is an oversight an we should switch to –sync-sgd for single GPU training by default.

From: Kelly Marchisio
Sent: Tuesday, March 19, 2019 8:26 PM
To: marian-nmt/marian
Cc: Subscribed
Subject: [marian-nmt/marian] confusion about optimizer delay - workingcorrectly? (Transformer Big) (#259)

I have two nearly-identical Transformer Big models that differ only in that optimizer-delay=1 in the first and optimizer-delay=4 in the second. Both have --mini-batch-fit, --maxi-batch 1000, and maxi-batch-sort: src .
Here are the first lines of training:
optimizer-delay=1 :
Ep. 1 : Up. 500 : Sen. 31905 : Cost 9.98 : Time 622.73s : 1409.48 words/s : L.r. 3.1250e-06
optimizer-delay=4
Ep. 1 : Up. 500 : Sen. 31905 : Cost 9.98 : Time 622.62s : 1409.73 words/s : L.r. 3.1250e-06
I'd have expected from this chain for Sen. in optimizer-delay=4 to be 4x what I saw in optimizer-delay=1. Furthermore, it appears my models are only seeing ~64 sentences per update, correct? This doesn't seem right --- right?

**Comment by kellymarchisio** [comment]
`--sync-sgd` is on for both of them, actually. But yes, I am on a single GPU

**Comment by kellymarchisio** [comment]
Full Config, if it's helpful. The other model differs only in optimizer-delay=1

[2019-03-16 17:30:15] [config] after-batches: 0
[2019-03-16 17:30:15] [config] after-epochs: 10
[2019-03-16 17:30:15] [config] allow-unk: false
[2019-03-16 17:30:15] [config] batch-flexible-lr: false
[2019-03-16 17:30:15] [config] batch-normal-words: 1920
[2019-03-16 17:30:15] [config] beam-size: 6
[2019-03-16 17:30:15] [config] best-deep: false
[2019-03-16 17:30:15] [config] clip-norm: 5
[2019-03-16 17:30:15] [config] cost-type: ce-mean-words
[2019-03-16 17:30:15] [config] cpu-threads: 0
[2019-03-16 17:30:15] [config] data-weighting-type: sentence
[2019-03-16 17:30:15] [config] dec-cell: gru
[2019-03-16 17:30:15] [config] dec-cell-base-depth: 2
[2019-03-16 17:30:15] [config] dec-cell-high-depth: 1
[2019-03-16 17:30:15] [config] dec-depth: 6
[2019-03-16 17:30:15] [config] devices:
[2019-03-16 17:30:15] [config]   - 3
[2019-03-16 17:30:15] [config] dim-emb: 1024
[2019-03-16 17:30:15] [config] dim-rnn: 1024
[2019-03-16 17:30:15] [config] dim-vocabs:
[2019-03-16 17:30:15] [config]   - 36000
[2019-03-16 17:30:15] [config]   - 36000
[2019-03-16 17:30:15] [config] disp-freq: 500
[2019-03-16 17:30:15] [config] dropout-rnn: 0
[2019-03-16 17:30:15] [config] dropout-src: 0
[2019-03-16 17:30:15] [config] dropout-trg: 0
[2019-03-16 17:30:15] [config] early-stopping: 5
[2019-03-16 17:30:15] [config] embedding-fix-src: false
[2019-03-16 17:30:15] [config] embedding-fix-trg: false
[2019-03-16 17:30:15] [config] embedding-normalization: false
[2019-03-16 17:30:15] [config] enc-cell: gru
[2019-03-16 17:30:15] [config] enc-cell-depth: 1
[2019-03-16 17:30:15] [config] enc-depth: 6
[2019-03-16 17:30:15] [config] enc-type: bidirectional
[2019-03-16 17:30:15] [config] exponential-smoothing: 0.0001
[2019-03-16 17:30:15] [config] grad-dropping-momentum: 0
[2019-03-16 17:30:15] [config] grad-dropping-rate: 0
[2019-03-16 17:30:15] [config] grad-dropping-warmup: 100
[2019-03-16 17:30:15] [config] guided-alignment-cost: ce
[2019-03-16 17:30:15] [config] guided-alignment-weight: 1
[2019-03-16 17:30:15] [config] ignore-model-config: false
[2019-03-16 17:30:15] [config] keep-best: true
[2019-03-16 17:30:15] [config] label-smoothing: 0.1
[2019-03-16 17:30:15] [config] layer-normalization: false
[2019-03-16 17:30:15] [config] learn-rate: 0.0001
[2019-03-16 17:30:15] [config] log: model/m-hyp-multi-lr1-big1/train.log
[2019-03-16 17:30:15] [config] log-level: info
[2019-03-16 17:30:15] [config] lr-decay: 0
[2019-03-16 17:30:15] [config] lr-decay-freq: 50000
[2019-03-16 17:30:15] [config] lr-decay-inv-sqrt: 16000
[2019-03-16 17:30:15] [config] lr-decay-repeat-warmup: false
[2019-03-16 17:30:15] [config] lr-decay-reset-optimizer: false
[2019-03-16 17:30:15] [config] lr-decay-start:
[2019-03-16 17:30:15] [config]   - 10
[2019-03-16 17:30:15] [config]   - 1
[2019-03-16 17:30:15] [config] lr-decay-strategy: epoch+stalled
[2019-03-16 17:30:15] [config] lr-report: true
[2019-03-16 17:30:15] [config] lr-warmup: 16000
[2019-03-16 17:30:15] [config] lr-warmup-at-reload: false
[2019-03-16 17:30:15] [config] lr-warmup-cycle: false
[2019-03-16 17:30:15] [config] lr-warmup-start-rate: 0
[2019-03-16 17:30:15] [config] max-length: 100
[2019-03-16 17:30:15] [config] max-length-crop: false
[2019-03-16 17:30:15] [config] maxi-batch: 1000
[2019-03-16 17:30:15] [config] maxi-batch-sort: src
[2019-03-16 17:30:15] [config] mini-batch: 1000
[2019-03-16 17:30:15] [config] mini-batch-fit: true
[2019-03-16 17:30:15] [config] mini-batch-fit-step: 10
[2019-03-16 17:30:15] [config] mini-batch-words: 0
[2019-03-16 17:30:15] [config] model: model/m-hyp-multi-lr1-big1/model.npz
[2019-03-16 17:30:15] [config] multi-node: false
[2019-03-16 17:30:15] [config] multi-node-overlap: true
[2019-03-16 17:30:15] [config] n-best: false
[2019-03-16 17:30:15] [config] no-reload: false
[2019-03-16 17:30:15] [config] no-restore-corpus: false
[2019-03-16 17:30:15] [config] no-shuffle: false
[2019-03-16 17:30:15] [config] normalize: 0.6
[2019-03-16 17:30:15] [config] optimizer: adam
[2019-03-16 17:30:15] [config] optimizer-delay: 4
[2019-03-16 17:30:15] [config] optimizer-params:
[2019-03-16 17:30:15] [config]   - 0.9
[2019-03-16 17:30:15] [config]   - 0.98
[2019-03-16 17:30:15] [config]   - 1e-09
[2019-03-16 17:30:15] [config] overwrite: true
[2019-03-16 17:30:15] [config] quiet: false
[2019-03-16 17:30:15] [config] quiet-translation: true
[2019-03-16 17:30:15] [config] relative-paths: false
[2019-03-16 17:30:15] [config] right-left: false
[2019-03-16 17:30:15] [config] save-freq: 5000
[2019-03-16 17:30:15] [config] seed: 1111
[2019-03-16 17:30:15] [config] skip: false
[2019-03-16 17:30:15] [config] sqlite: ""
[2019-03-16 17:30:15] [config] sqlite-drop: false
[2019-03-16 17:30:15] [config] sync-sgd: true
[2019-03-16 17:30:15] [config] tempdir: /home/kmarc/experiment/jhu-wmt18/tmp
[2019-03-16 17:30:15] [config] tied-embeddings: false
[2019-03-16 17:30:15] [config] tied-embeddings-all: true
[2019-03-16 17:30:15] [config] tied-embeddings-src: false
[2019-03-16 17:30:15] [config] train-sets:
[2019-03-16 17:30:15] [config]   - data/wmt17x2_back_p8Me2M4.bpe.en
[2019-03-16 17:30:15] [config]   - data/wmt17x2_back_p8Me2M4.bpe.de
[2019-03-16 17:30:15] [config] transformer-dim-ffn: 4096
[2019-03-16 17:30:15] [config] transformer-dropout: 0.1
[2019-03-16 17:30:15] [config] transformer-dropout-attention: 0.1
[2019-03-16 17:30:15] [config] transformer-dropout-ffn: 0.1
[2019-03-16 17:30:15] [config] transformer-ffn-activation: swish
[2019-03-16 17:30:15] [config] transformer-heads: 8
[2019-03-16 17:30:15] [config] transformer-postprocess: dan
[2019-03-16 17:30:15] [config] transformer-postprocess-emb: d
[2019-03-16 17:30:15] [config] transformer-preprocess: ""
[2019-03-16 17:30:15] [config] type: transformer
[2019-03-16 17:30:15] [config] valid-freq: 5000
[2019-03-16 17:30:15] [config] valid-log: model/m-hyp-multi-lr1-big1/valid.log
[2019-03-16 17:30:15] [config] valid-max-length: 1000
[2019-03-16 17:30:15] [config] valid-metrics:
[2019-03-16 17:30:15] [config]   - ce-mean-words
[2019-03-16 17:30:15] [config]   - perplexity
[2019-03-16 17:30:15] [config] valid-mini-batch: 16
[2019-03-16 17:30:15] [config] valid-script-path: ./scripts/validate.de.sh
[2019-03-16 17:30:15] [config] valid-sets:
[2019-03-16 17:30:15] [config]   - data/valid.bpe.en
[2019-03-16 17:30:15] [config]   - data/valid.bpe.de
[2019-03-16 17:30:15] [config] valid-translation-output: model/m-hyp-multi-lr1-big1/valid.bpe.en.output
[2019-03-16 17:30:15] [config] version: v1.4.0+f429d4a8
[2019-03-16 17:30:15] [config] vocabs:
[2019-03-16 17:30:15] [config]   - data/vocab-corpus17back.ende.yml
[2019-03-16 17:30:15] [config]   - data/vocab-corpus17back.ende.yml
[2019-03-16 17:30:15] [config] workspace: 6500

**Comment by emjotde** [comment]
Oh weird. I am sure I used this a few days ago. I will check tomorrow and report back.

**Comment by cepin19** [comment]
I had a similar problem few months ago - optimizer delay didn't seem to work on a single GPU, but since about version 1.7.something it looks like it works properly, so maybe try updating Marian.

**Comment by kellymarchisio** [comment]
ah, thank you, @cepin19.  @Marcin, if you agree that this is malfunctioning, I'll kill my jobs and try updating Marian again. (@ykl7 on [this thread](https://github.com/marian-nmt/marian/issues/250) is my teammate and we haven't gotten 1.7 working yet, but I'll try the suggestions mentioned in the last message.

**Comment by emjotde** [comment]
Ah, old Marian version, ancient actually... 1.4. OK, I was just going to answer in a very confused way that it works perfectly fine for me. As a side note, posting issues concerning old versions is not the best idea :) We kind of assume you use the current release.

**Comment by kellymarchisio** [comment]
Ah yeah, sorry about that - slipped my mind. We finally got 1.7 working late last night so I expect it'll work there. Thanks for looking.

**Comment by ZJaume** [comment]
Hi, I used `optimizer-delay 4` on a single gpu and noticed that batch size was the same that with delay 1, because `sync-sgd` was set to `false` by default. Maybe `sync-sgd` needs to be `true` by default with a single device, or at least add to `optimizer-delay` help that it has to be used with `sync-sgd`.

I'm using Marian 1.8.27

**Comment by frankseide** [comment]
Please use `--sync-sgd` for now even with a single GPU. It is the most complete implementation. Most other SGD implementations are sort of legacy.

**Comment by ZJaume** [comment]
So, why `sync-sgd` is not enabled by default for single-gpu?
